### ISSUE TITLE
Decide on whether to include failed transactions in blocks

### ISSUE THREAD

**Issue by geigerzaehler** [oscoin/oscoin#500]
Transactions might fail to apply to the state when a node tries to include them in a block. We need to decide whether to include failed transactions in blocks or not.

An example of a failure is a transaction that transfers 10 coins from account A to B but is applied to a state where account A has less than 10 coins.

### Considerations

There are two points that we need to consider when making this decision

#### Feedback for clients

Clients that submit a transaction want to know if the transactions was successfully If we don’t provide a notification mechanism for failed transactions a client cannot decide whether the transaction has not been included yet and it should wait more or whether the transaction failed and it should notify the user. Note that a feedback mechanism does not require the transaction to be included in the blockchain. Transaction errors might be gossiped separately. In this case the answer is un-authorative however.

#### Replay attacks

It is easy to conceive of a replay attack if transactions are not included in the blockchain. Assume that A buys something from B and creates a transaction that sends the required amount from A to B. But by the time the transaction arrives in the network A has insufficient funds. Since B does not see a confirmation of the transaction it cancels the transaction. After some time B has become malicious and A has received some money. If A has not issued a transaction in between B can replay the original transaction and transfer the funds.

### Comparison with other Blockchains

In Bitcoin a transaction can fail if a transaction input has already been spent. Failing transactions are not included in blocks and nodes will not relay them. 

In Ethereum failed transactions are included in the blockchain. (See for example [this transaction](https://ropsten.etherscan.io/tx/0x67a5f6442f49a5da6ff8682250a8eef899d9dc0c5adf20b683709433902b5956)).

### Current implementation

The current implementation includes failed transactions in the block. Transaction failures are present in the type of `Evaluator state tx output :: tx -> state -> Either EvalError (output, state)`. The receipt generated by a transaction in a block has the field `receiptTxOutput :: Either EvalError output` which tells us whether a transaction failed or was successful. The function `evalBlock` generates these receipts by applying an `Evaluator`.

**Comment by kim** [comment]
What are reasons **not** to include failed transactions in the block?

**Comment by geigerzaehler** [comment]
> What are reasons **not** to include failed transactions in the block?

The only arguments I could think of are block size and wasted computation. I added those to the description.

**Comment by cloudhead** [comment]
Another reason for including them is that miners can't get the tx fee from them if they aren't included. This *could* be a potentialy DoS vector, eg. submitting txs that are expensive but fail right at the end.

**Comment by adinapoli-mndc** [comment]
Thinking out loud here (mostly as I come from a UTxO background where some of these things are a bit more deterministic). The TL;DR is that I think including failed transaction in blocks sounds sensible based on what we are describing so far. However, there are a bunch of open questions we need to think about (which could just as well be orthogonal to this particular issue):

* **Client feedback**: it seems to me that some of those replay attacks / feedback concerns would go away if we introduce a TTL (_time-to-live_) for a transaction. In systems like Cardano this could be the number of slots up until a tx is considered valid, but for us could be a `NominalDiffTime` or a timestamp of the time it expires or some other counter. The point being it would be easier for clients to understand whether or not a tx is failed if we eliminate the guesswork and we store the correct information in our single-source-of-truth (i.e. the blockchain). This also allows node mempools to prune expired transactions without evaluating them, which can be costly.

* **Tx fees**: One of the reasons why you pay fees for a transaction is also as a self-limiting mechanism to avoid flooding the network with failable/small txs. In account-based models (like ETH) it seems like you _always_ pay a fee, regardless whether or not your Tx will fail or not. I guess this is a bit unfair on the user,but I don't see immediately how you could prevent adversarial behaviour otherwise;

On the last point, it's perhaps surprising to find failed transactions included in the _main chain_ (like the one linked by Thomas) as opposed as just in mined blocks gossiped around. However, it seems like in ETH they had to choice, as each Tx generates fees (regardless if it's invalid or not) and thus contributes to modify the final account's balance of the user. Furthermore, this makes sense in case of failures arising from smart contract scripts, as they would be wasting resources (i.e. gas) and as such they have to pay for the cost of executing the script.

**Comment by cloudhead** [comment]
Good points - one issue with the TTL idea is that if it's too short, valid transactions may expire before they are mined, and if it's too long, clients have to wait longer than if they just saw the failed tx in a block.

**Comment by adinapoli-mndc** [comment]
> Good points - one issue with the TTL idea is that if it's too short, valid transactions may expire before they are mined, and if it's too long, clients have to wait longer than if they just saw the failed tx in a block.

Right, but this would make the decision process deterministic -- if a client sees a failed tx in a block (assuming of course that's not persisted on main chain but it's in the mempool) how could the user know for sure this transaction won't become valid at some point in the future?

In other terms: if a failed tx is on the main chain (with enough number of confirmations and/or in a non-mutable part of the chain) by all means you don't need the TTL, you can state unambiguously that the Tx failed, but if the tx is in one of the volatile blocks, that's tricker to answer.

**Comment by cloudhead** [comment]
> Right, but this would make the decision process deterministic -- if a client sees a failed tx in a block (assuming of course that's not persisted on main chain but it's in the mempool) how could the user know for sure this transaction won't become valid at some point in the future?
> 
If the tx is in a block, then it shouldn't be in a mempool, modulo propagation time of that block. The thing is clients only ever see the main chain, they don't see any forks, so if it's in the main chain and failed, they will have to retry it.

**Comment by adinapoli-mndc** [comment]
> If the tx is in a block, then it shouldn't be in a mempool, modulo propagation time of that block.

Right 🤦‍♂️

**Comment by cloudhead** [comment]
Can we finalize this? I think most arguments are in favor of including failed txs.

**Comment by geigerzaehler** [comment]
> Can we finalize this? I think most arguments are in favor of including failed txs.

I’m in favor of including failed transactions.

**Comment by adinapoli-mndc** [comment]
I would say to crack on with the plan (i.e. _include failed transactions_) and review (if needed) if this choice proved to be unergonomic once we spin up our devnet/testnet, once we have a chance to gather real usability information.
